crypto/aes/asm/aes-cfb-avx512.pl

```diff
@@ -0,0 +1,921 @@
+    vaesenclast %zmm29,$temp_8x, $temp_8x
+    vaesenclast %zmm29,$temp_12x,$temp_12x
+    vaesenclast %zmm29,$temp_16x,$temp_16x
+    jmp ${label_prefix}_end
+
+.balign 32
+${label_prefix}_256:
+
+    vaesenc %zmm29,$temp_4x, $temp_4x
+    vaesenc %zmm29,$temp_8x, $temp_8x
+    vaesenc %zmm29,$temp_12x,$temp_12x
+    vaesenc %zmm29,$temp_16x,$temp_16x
+
+    vaesenc %zmm30,$temp_4x, $temp_4x
+    vaesenc %zmm30,$temp_8x, $temp_8x
+    vaesenc %zmm30,$temp_12x,$temp_12x
+    vaesenc %zmm30,$temp_16x,$temp_16x
+
+    vaesenclast %zmm31,$temp_4x, $temp_4x
+    vaesenclast %zmm31,$temp_8x, $temp_8x
+    vaesenclast %zmm31,$temp_12x,$temp_12x
+    vaesenclast %zmm31,$temp_16x,$temp_16x
+
+.balign 32
+${label_prefix}_end:
+___
+}
+
+#################################################################
+# Signature:
+#
+# void aes_cfb128_vaes_enc(
+#     const unsigned char *in,
+#     unsigned char *out,
+#     size_t len,
+#     const AES_KEY *ks,
+#     const unsigned char ivec[16],
+#     /*in-out*/ ossl_ssize_t *num);
+#
+# Preconditions:
+# - all pointers are valid (not NULL...)
+# - AES key schedule and rounds in `ks` are precomputed
+#
+# Invariants:
+# - `*num` is between 0 and 15 (inclusive)
+#################################################################
+
+$code.=<<___;
+.globl   aes_cfb128_vaes_enc
+.type    aes_cfb128_vaes_enc,\@function,6
+.balign  64
+aes_cfb128_vaes_enc:
+.cfi_startproc
+    endbranch
+___
+
+$inp="%rdi";          # arg0
+$out="%rsi";          # arg1
+$len="%rdx";          # arg2
+
+$key_original="%rcx"; # arg3
+$key_backup="%r10";
+
+$ivp="%r8";           # arg4
+$nump="%r9";          # arg5
+
+$num="%r11";
+$left="%rcx";
+$mask="%rax";
+
+$rounds="%r11d";
+
+$temp="%xmm2";
+$plain="%xmm3";
+
+$code.=<<___;
+
+    mov ($nump),$num                 # $num is the current byte index in the first partial block
+                                     # $num belongs to 0..15; non-zero means a partial first block
+
+    test $len,$len                   # return early if $len==0, unlikely to occur
+    jz .Laes_cfb128_vaes_enc_done
+
+    test $num,$num                   # check if the first block is partial
+    jz .Laes_cfb128_enc_mid          # if not, jump to processing full blocks
+
+###########################################################
+# first partial block pre-processing
+###########################################################
+
+    mov $key_original,$key_backup    # make room for variable shl with cl
+
+    mov \$0x10,$left                 # first block is partial
+    sub $num,$left                   # calculate how many bytes $left to process in the block
+    cmp $len,$left                   #
+    cmova $len,$left                 # $left = min(16-$num,$len)
+
+    mov \$1,$mask                    # build a mask with the least significant $left bits set
+    shlq %cl,$mask                   # $left is left shift counter
+    dec $mask                        # $mask is 2^$left-1
+    kmovq $mask,%k1
```

Comment:
I assume this doesnt need to be reset since anything that uses a k1 would normally set it first?

Reply:
They are scratch registers/caller save/volatile on 32-bit/64-bit Linux/Windows.